src/librustc_trait_selection/traits/select/confirmation.rs

```diff
@@ -808,3 +947,50 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
         Ok(ImplSourceBuiltinData { nested })
     }
 }
+
+struct ObjectAssociatedTypeNormalizer<'a, 'tcx> {
+    infcx: &'a infer::InferCtxt<'a, 'tcx>,
+    object_ty: Ty<'tcx>,
+    object_bounds: &'a [ty::ExistentialProjection<'tcx>],
+    param_env: ty::ParamEnv<'tcx>,
+    cause: &'a ObligationCause<'tcx>,
+    nested: &'a mut Vec<PredicateObligation<'tcx>>,
+}
+
+impl<'tcx> TypeFolder<'tcx> for ObjectAssociatedTypeNormalizer<'_, 'tcx> {
```

Comment:
This is a bit hacky. I'm going to try making a change to how project works for trait objects for this and see if that's better,